pkg/controller/endpoint/endpoints_controller_test.go

```diff
@@ -1990,6 +1990,200 @@ func TestEndpointPortFromServicePort(t *testing.T) {
+			blockNextAction <- struct{}{}
+		}
+
+		res.WriteHeader(http.StatusOK)
+		res.Write([]byte(runtime.EncodeOrDie(clientscheme.Codecs.LegacyCodec(v1.SchemeGroupVersion), &v1.Endpoints{})))
+	}
+
+	mux.HandleFunc("/api/v1/namespaces/"+ns+"/endpoints", handlerFunc)
+	mux.HandleFunc("/api/v1/namespaces/"+ns+"/endpoints/", handlerFunc)
+	mux.HandleFunc("/api/v1/namespaces/"+ns+"/events", func(res http.ResponseWriter, req *http.Request) {})
+	mux.HandleFunc("/", func(res http.ResponseWriter, req *http.Request) {
+		t.Errorf("unexpected request: %v", req.RequestURI)
+		http.Error(res, "", http.StatusNotFound)
+	})
+	testServer := httptest.NewServer(mux)
+
+	defer testServer.Close()
+	endpoints = newController(testServer.URL, 0*time.Second)
+	endpointsStore = endpoints.endpointsStore
+
+	addPods(endpoints.podStore, ns, 1, 1, 0, false)
+	go func() { endpoints.Run(1, stopChan) }()
+
+	svc := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: ns},
+		Spec: v1.ServiceSpec{
+			Selector:  map[string]string{"foo": "bar"},
+			ClusterIP: "None",
+			Ports:     nil,
+		},
+	}
+
+	endpoints.serviceStore.Add(svc)
+	endpoints.onServiceUpdate(svc)
+	timer := time.NewTimer(1 * time.Second)
+	select {
+	// BlockcNextAction should eventually unblock once server gets endpoint request.
+	case <-timer.C:
+		t.Errorf("Service Add should have caused a request to be sent to the test server")
+	case <-blockNextAction:
+	}
+
+	endpoints.serviceStore.Delete(svc)
+	endpoints.onServiceDelete(svc)
+	timer = time.NewTimer(1 * time.Second)
+	select {
+	// BlockcNextAction should eventually unblock once server gets endpoint request.
+	case <-timer.C:
+		t.Errorf("Service Delete should have caused a request to be sent to the test server")
+	case <-blockNextAction:
+	}
+
+	// if endpoints cache has not updated becfore service update is registered
+	// Services add will not trigger a Create endpoint request.
+	endpoints.serviceStore.Add(svc)
+	endpoints.onServiceUpdate(svc)
+
+	// Ensure service add has been processed
+	time.Sleep(1 * time.Second)
+	close(blockDelete)
+
+	timer = time.NewTimer(1 * time.Second)
+	select {
+	// Eventually the endpoint delete should cause the controller to recreate the endpoint
+	case <-timer.C:
+		t.Errorf("Endpoint should have been recreated")
+	case <-blockNextAction:
+	}
+
+	close(blockNextAction)
+	close(stopChan)
+}
+
+func TestEndpointsInvalidStates(t *testing.T) {
+	ns := metav1.NamespaceDefault
+	testServer, _ := makeTestServer(t, ns)
+	defer testServer.Close()
+	endpoints := newController(testServer.URL, 0)
+	tracker := endpoints.endpointsTracker
+	store := endpoints.endpointsStore
+	ep1 := &v1.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            "foo",
+			Namespace:       ns,
+			ResourceVersion: "rv1",
+		},
+	}
+
+	// Test Unexpected and Expected Adds
+	if endpoints.queue.Len() != 0 {
+		t.Errorf("Queue should be empty")
+	}
+	store.Add(ep1)
+	endpoints.onEndpointsUpdate(ep1)
+	if endpoints.queue.Len() != 1 {
```

Comment:
can we have a race here if we assert before the queue is populated?

Reply:
This test does not try to replicate the race, but just makes sure the service is added to the queue to be processed when an unexpected endpoints event comes through. `TestMultipleServiceChanges` covers the race condition.